genesis/sensors/contact_force.py

```diff
@@ -0,0 +1,306 @@
+    output: ti.types.ndarray(),
+):
+    for i_b, i_c, i_s in ti.ndrange(output.shape[0], link_a.shape[-1], sensors_link_idx.shape[-1]):
+        contact_data_link_a = link_a[i_b, i_c]
+        contact_data_link_b = link_b[i_b, i_c]
+        if contact_data_link_a == sensors_link_idx[i_s] or contact_data_link_b == sensors_link_idx[i_s]:
+            j_s = i_s * 3  # per-sensor output dimension is 3
+
+            quat_a = ti.Vector.zero(ti.f32, 4)
+            quat_b = ti.Vector.zero(ti.f32, 4)
+            for j in ti.static(range(4)):
+                quat_a[j] = links_quat[i_b, contact_data_link_a, j]
+                quat_b[j] = links_quat[i_b, contact_data_link_b, j]
+
+            force_vec = ti.Vector.zero(ti.f32, 3)
+            for j in ti.static(range(3)):
+                force_vec[j] = contact_forces[i_b, i_c, j]
+
+            force_a = ti_inv_transform_by_quat(-force_vec, quat_a)
+            force_b = ti_inv_transform_by_quat(force_vec, quat_b)
+
+            if contact_data_link_a == sensors_link_idx[i_s]:
+                for j in ti.static(range(3)):
+                    output[i_b, j_s + j] += force_a[j]
+            if contact_data_link_b == sensors_link_idx[i_s]:
+                for j in ti.static(range(3)):
+                    output[i_b, j_s + j] += force_b[j]
+
+
+class ContactSensorOptions(RigidSensorOptionsMixin, SensorOptions):
+    """
+    Sensor that returns bool based on whether associated RigidLink is in contact.
+
+    Parameters
+    ----------
+    entity_idx : int
+        The global entity index of the RigidEntity to which this sensor is attached.
+    link_idx_local : int, optional
+        The local index of the RigidLink of the RigidEntity to which this sensor is attached.
+    delay : float
+        The delay in seconds before the sensor data is read.
+    update_ground_truth_only : bool
+        If True, the sensor will only update the ground truth data, and not the measured data.
+    """
+
+
+@dataclass
+class ContactSensorMetadata(SharedSensorMetadata):
+    """
+    Metadata for all rigid contact sensors.
+    """
+
+    solver: RigidSolver | None = None
+    expanded_links_idx: torch.Tensor = field(
+        default_factory=lambda: torch.tensor([], dtype=gs.tc_float, device=gs.device)
+    )
+
+
+@register_sensor(ContactSensorOptions, ContactSensorMetadata)
+@ti.data_oriented
+class ContactSensor(Sensor):
+    """
+    Sensor that returns bool based on whether associated RigidLink is in contact.
+    """
+
+    def build(self):
+        if self._shared_metadata.solver is None:
+            self._shared_metadata.solver = self._manager._sim.rigid_solver
+
+        self._shared_metadata.expanded_links_idx = torch.cat(
+            [
+                self._shared_metadata.expanded_links_idx,
+                torch.tensor(
+                    [self._options.entity_idx + self._options.link_idx_local], dtype=gs.tc_int, device=gs.device
+                )
+                .unsqueeze(0)
```

Comment:
Avoid unsqueeze when it could be defined unsqueezed directly:
```python
                torch.tensor(
                    [[self._options.entity_idx + self._options.link_idx_local]], dtype=gs.tc_int, device=gs.device
                )
```

But here you should just use `torch.full` because efficiency is not a concern during build.